### ISSUE TITLE
Non-static API for play.libs.Json (Java)

### ISSUE THREAD

**Issue by gmethvin** [playframework/playframework#6332]
We can make `play.libs.Json` an injectable class with instance methods. It should be possible to inject `Json`, `ObjectMapper`, `ObjectReader` and `ObjectWriter`. Maybe the static methods could stay in undeprecated form but only delegate to a global default instance (not the application).

**Comment by benmccann** [comment]
Most of my controller methods do something like:

    return Json.toJson(pojo);

I agree with the goal of preferring injection over static methods. We should try to at the same time not add too much extra code for users. In this instance construction the controller would become more cumbersome since every single controller I have would need to inject an instance of `Json`.

Really what I'd like to be able to do is simply `return pojo;`. Then the framework can handle converting the object to JSON for me and can use an injected serializer without me having to do the work of doing the serialization.

We would probably have to support the ability to serialize to both JSON and XML. There's a few options I can think of here:

* Specified in the URL. E.g. some API designers advocate using a .json or .xml extension to the endpoint URL to indicate what the URL will return
* An annotation on a per-controller/action basis like Dropwizard uses: `@Produces(MediaType.APPLICATION_JSON)`
* Some additional markup in the routes file to specify XML vs JSON

**Comment by gmethvin** [comment]
Hi @benmccann,

I think `play.libs.Json` would be one place where we'd keep the static API available (there may be other places where it's justified, not sure yet).

I assume you mean `return ok(pojo)` or similar? That might be interesting, though if we did that we'd probably also need to provide some generic mechanism for mapping pojos to responses. Actually I think this should be already possible in Scala by defining a custom `Writeable[Pojo]` but it's not obvious or documented anywhere that I know of.

**Comment by benmccann** [comment]
Yeah, I guess I forgot the call to `ok`. I don't think mapping a pojo to a response would be too hard. Just call `Json.toJson(pojo)`.

**Comment by schmitch** [comment]
Actually I don't think its a good idea to have a `Json.toJson(pojo)`.
First of, if we provide that, people sooner or later want a way to have the content rendered based on their `Accept` Header.
Second it's not a good idea to directly serialize List<Pojo> into Json, it's mostly good to wrap that.

Also for Scala there is also no way to do that as far as I know but actually I think for a Framework like Play! it shouldn't do it since it's really not "hard" to do it on your own and I think that when we do it, we should actually do it in a extra project and have a "special" Controller for that.

**Comment by gmethvin** [comment]
Well, I'm not sure it makes sense to implement it, but if we did it would make sense for it to be a generic mechanism. The problem (as you pointed out @schmitch) is that you can't know if the arbitrary type you pass will serialize well to JSON and/or XML.

In Scala, actions can be created by arbitrary functions that return `Action[A]`, so you can easily create your own custom action function that accepts a block from `Request => T` (or something like that), and transforms the `T` into a result. You can also use type classes to ensure that there are serializers available for `T` (e.g. `Writes[T]` for JSON), so it's a bit "safer". It's not directly provided by the framework but it's easy enough to set up yourself that it doesn't matter.

**Comment by PromanSEW** [comment]
See also #6967
